Revamp port forwarding to support UDP

balajiv113 opened the pull request:
Fixes #366 
Discussion - #2402 

Revamps port forwarding to use existing GRPC communication.

**Advantages**
- No SSH daemon (See Notes below)
- No Subprocess for SSH
- Faster (See performance below)
- Ease of support for newer protocol (We do support UDP as well with this implementation)

**Todo**
- [x] GRPC based tunnels
- [x] TCP support
- [x] UDP support
- [x] Config to enable / disable this new port forwarding model
- [x] Testing on MacOS
- [ ] Testing on Linux
- [ ] Testing on Windows

**Performance**
GRPC TCP - ~3.80 Gbits/sec
GRPC TCP Reverse - ~4.77 Gbits/sec
SSH TCP - ~3.38 Gbits/sec
SSH TCP Reverse - ~3.08 Gbits/sec

**Notes**
- We will support only TCP and UDP via this approach. Unix and unixgram support socket to be added later as a separate PR
- We support forwarding from host (Listen) -> guest (Dial). We don't support reverse forwarding host (Dial) -> guest (Listen) as of this PR. But can be extended and support it in a separate PR

alexandear commented on pkg/guestagent/api/guestservice.proto:
```diff
@@ -22,11 +24,20 @@ message Event {
 message IPPort {
-  string ip = 1;
-  int32 port = 2;
+  string protocol = 1;
+  string ip = 2;
+  int32 port = 3;
```
Perhaps this is a breaking change. According to https://protobuf.dev/programming-guides/proto3/#assigning  "This number cannot be changed once your message type is in use"

Better:

```suggestion
  string ip = 1;
  int32 port = 2;
  string protocol = 3;
```

balajiv113 commented:
Converting to draft as it needs some more changes related to handling of ports on host side

balajiv113 commented:
@lima-vm/maintainers 
Do we need to back this feature on with a config in lima yaml ?? 

Just thinking because we are looking at this as a drop-in replacement. If we went with config we need to mark this experimental and later bring it to stable / default. Also it will pollute the yaml as well

One another idea is to back this with a ENV / start option

AkihiroSuda commented:
> ENV

👍

balajiv113 commented:
Added a env and updated docs as well for this feature.

@lima-vm/maintainers 
We are good to review this PR.

I have done decent amount of testing in macOS and a very basic test in linux. For performance tried with iperf3 for both tcp and udp.
Need some help to test in windows (Not sure if we even support port forwarding here, i think wsl takes care of it)

AkihiroSuda commented on .github/workflows/test.yml:
```diff
@@ -16,6 +16,7 @@ on:
     - "**.md"
 env:
   LIMACTL_CREATE_ARGS: ""
+  LIMA_GRPC_PORT_FORWARDER: true
```
Can we just enable this by default?

balajiv113 replied:
yes we can. I did this just to be on safer side and not break any exiting feature.

If you feel we can make this default let me know i will update it

AkihiroSuda replied:
I guess this can be enabled by default

balajiv113 replied:
Done.

AkihiroSuda commented on pkg/guestagent/api/guestservice.proto:
```diff
@@ -22,11 +24,20 @@ message Event {
+
+message TunnelMessage {
+  string id = 1;
+  string protocol = 2;
```
Could you explain the allowed values in comment lines?

balajiv113 replied:
Done.

AkihiroSuda commented on pkg/hostagent/hostagent.go:
```diff
@@ -164,6 +166,11 @@ func New(instName string, stdout io.Writer, signalCh chan os.Signal, opts ...Opt
 	limayaml.FillPortForwardDefaults(&rule, inst.Dir)
 	rules = append(rules, rule)
 
+	env := os.Getenv("LIMA_GRPC_PORT_FORWARDER")
```
This should be parsed with `strconv.ParseBool`

balajiv113 replied:
Done.

AkihiroSuda commented on website/content/en/docs/Config/Port/_index.md:
```diff
@@ -0,0 +1,35 @@
+
+#### Advantages
+- Supports both TCP and UDP based port forwarding
+- Performs faster compared to SSH based forwarding
```
Can we show some benchmark result here?

balajiv113 replied:
Done.

AkihiroSuda commented on website/content/en/docs/Config/Port/_index.md:
```diff
@@ -0,0 +1,35 @@
+weight: 50
+---
+
+Lima supports automatic port-forwarding of ports from guest to host.
```
Can we explain that this only applies to localhost ports

balajiv113 replied:
Done.

AkihiroSuda commented on pkg/guestagent/procnettcp/procnettcp.go:
```diff
@@ -15,14 +15,17 @@ type Kind = string
 	TCP6 Kind = "tcp6"
-	// TODO: "udp", "udp6", "udplite", "udplite6".
+	UDP  Kind = "udp"
+	// TODO: "udplite", "udplite6".
```
what about udp6?

balajiv113 replied:
My bad, missed it. Will enable for udp6 as well.

For udplite, updlite6 we can enable in a separate PR. It should work as is mostly.

balajiv113 replied:
Done.

AkihiroSuda reviewed (dismissed):
Thanks

AkihiroSuda commented on website/content/en/docs/Config/Port/_index.md:
```diff
@@ -0,0 +1,65 @@
+> **Warning**
+> This mode is experimental
+
+| ⚡ Requirement | Lima >= 0.23 + |
```
```suggestion
| ⚡ Requirement | Lima >= 0.23 |
```

balajiv113 replied:
Done

AkihiroSuda commented on website/content/en/docs/Config/Port/_index.md:
```diff
@@ -0,0 +1,65 @@
+### Using GRPC (Default)
+
+> **Warning**
+> This mode is experimental
```
Does this apply to TCP too?

balajiv113 replied:
Yes correct. The GRPC tunnel itself new right so it applies for both tcp and usp

AkihiroSuda commented:
@lima-vm/maintainers Can we merge this?

moenodedev commented:
@jandubois @afbjorklund Do you need more testers?

jandubois commented:
I'm sorry, I won't have time to test/review this until September. 😞

suhailskhan commented:
@balajiv113 I have been testing UDP port forwarding with some Docker Compose applications, and I may have found an issue.

After I first run `docker compose up`, any UDP ports that I choose to expose are reachable, as expected. However, I find that after I restart the application, those ports are not reachable anymore. This does not happen with TCP port forwarding. TCP ports that I expose in the container(s) are reachable even after restarting. The only way I was able to make the UDP ports for my Compose apps reachable again was to restart the Lima VM. However, the problem reappears when I restart the Compose app.

To test, I built Lima from your most recent commit in this PR at the time of writing this comment. I created a VM using the Docker template with the following command: `limactl create --vm-type vz template://docker --name=default`. Below is the `docker-compose.yml` for a simple Docker Compose application that demonstrates the issue:
```yaml
services:
  iperf3-server:
    image: networkstatic/iperf3
    command: -s
    ports:
      - "5201:5201/udp"
      - "5201:5201/tcp"
```
To verify whether or not the `iperf3-server` container is listening on TCP port 5201 and UDP port 5201, I used `sudo lsof -i TCP:5201` and `sudo lsof -i UDP:5201`, respectively. If it is listening on a port, the command should have output. If there is no output, then that means the service is not listening on the port. When I run the Compose app for the first time, I get output from both commands. After restarting the app, I only get output from the `lsof` command for the TCP port.

Although I was primarily using Docker Compose in my tests, it seems to affect Docker containers in general. If I start a container with a certain UDP port forwarded, then stop it, then that UDP port is not able to be exposed in subsequently created containers, until the Lima VM is restarted.

Would be curious to see if you or anyone else can reproduce this.

I am testing this on macOS Sonoma 14.5 on an Intel Mac. I am also able to test on Sonoma machine with Apple silicon.

balajiv113 commented:
@suhailskhan Thanks for trying it out. 

Yes it is indeed a issue. It happens due to conn cache on guest side. On closing of listener its not cleared properly
Will fix and let you know

balajiv113 commented:
@suhailskhan Fixed it. It should work now

suhailskhan commented:
@balajiv113 Pulled, rebuilt, and tested. Works indeed!

AkihiroSuda commented on website/content/en/docs/config/Port/_index.md:
```diff
@@ -0,0 +1,65 @@
+> **Warning**
+> This mode is experimental
+
+| ⚡ Requirement | Lima >= 0.23 |
```
```suggestion
| ⚡ Requirement | Lima >= 1.0 |
```

balajiv113 replied:
Done

AkihiroSuda commented on website/content/en/docs/config/Port/_index.md:
```diff
@@ -0,0 +1,65 @@
+- Doesn't support UDP based port forwarding
+- Spans child process on host for running SSH master.
+
+### Using GRPC (Default)
```
```suggestion
### Using GRPC (Default since Lima v1.0)
```

balajiv113 replied:
Done

AkihiroSuda commented on website/content/en/docs/config/Port/_index.md:
```diff
@@ -0,0 +1,65 @@
+
+### Using SSH
+
+SSH based port forwarding is the default and current model that is supported in lima.
```
```suggestion
SSH based port forwarding is the default and current model that is supported in Lima prior to v1.0
```

balajiv113 replied:
Done

AkihiroSuda commented:
Lint is failing

balajiv113 commented:
@AkihiroSuda Done. All Green

AkihiroSuda reviewed (approved):
Thanks
